### PR TITLE
Retain sealing and opening key

### PR DESCRIPTION
**Pull request by stammw**
Here is a pretty trivial solution for #17. Is it what you expected ?

Similarly to #92, built keys and the data used to build them could be constructed wrongly by using a bad `CryptoContext { ... }`. I propose to move key and `CrytpoState` building logic into a `CryptoContext::new()` method to prevent this from happening.

Everything would be easier with all the key types implementing `Clone`. I'll see what can be done on this side. But I wonder why it is not already the case?

### CODE REVIEW

File: quinn-proto/src/crypto.rs [hunk 1]
```diff
@@ -270,20 +273,25 @@ impl Crypto {
         self.write_nonce(&state, packet, nonce);
         let payload_len = payload.len();
 
-        let key = aead::OpeningKey::new(cipher, &state.key).unwrap();
         aead::open_in_place(&key, &*nonce, header, 0, payload.as_mut()).map_err(|_| ())?;
         payload.split_off(payload_len - cipher.tag_len());
         Ok(())
     }
 
     pub fn update(&self, side: Side) -> Crypto {
```
Comment (stammw) [inline quinn-proto/src/crypto.rs]: Couldn't `CryptoContext::local` and `CryptoContext::remote` just be swapped in `CryptoContext::update` `CryptoState::update` use the same `cipher` and `digest` [here](https://github.com/djc/quinn/blob/f9a6935b2ffcda1f08e600b5ab4b87668965a5fc/quinn-proto/src/crypto.rs#L384) ?

File: quinn-proto/src/crypto.rs [hunk 2]
```diff
@@ -390,12 +398,26 @@ pub struct ZeroRttCrypto {
...
+    fn clone(&self) -> CryptoContext {
+        Self {
+            sealing_key: aead::SealingKey::new(self.cipher, &self.local.key.clone()).unwrap(),
+            opening_key: aead::OpeningKey::new(self.cipher, &self.remote.key).unwrap(),
+            local: self.local.clone(),
+            remote: self.remote.clone(),
+            digest: self.digest,
+            cipher: self.cipher,
+        }
+    }
```
Comment (Ralith) [inline quinn-proto/src/crypto.rs]: I don't think we actually clone this structure ever, so we might as well remove the impl.

File: quinn-proto/src/crypto.rs (hunk 1 above)
Comment (Ralith) [reply quinn-proto/src/crypto.rs]: I don't understand this question; can you rephrase?

File: quinn-proto/src/crypto.rs [hunk 3]
```diff
@@ -390,12 +398,26 @@ pub struct ZeroRttCrypto {
     cipher: &'static aead::Algorithm,
 }
 
-#[derive(Clone)]
 pub struct CryptoContext {
     local: CryptoState,
     remote: CryptoState,
     digest: &'static digest::Algorithm,
     cipher: &'static aead::Algorithm,
```
Comment (Ralith) [inline quinn-proto/src/crypto.rs]: *ring* keys store their `&'static aead::Algorithm` internally, so we can remove the `cipher` member.

File: quinn-proto/src/crypto.rs [hunk 4]
```diff
@@ -390,12 +398,26 @@ pub struct ZeroRttCrypto {
...
 }
 
-#[derive(Clone)]
 pub struct CryptoContext {
     local: CryptoState,
     remote: CryptoState,
```
Comment (Ralith) [inline quinn-proto/src/crypto.rs]: As in the previous PR, I'd like to avoid storing the key bytes redundantly. This will probably require some refactoring, perhaps including doing away with `CryptoState`.

File: quinn-proto/src/crypto.rs (hunk 1 above)
Comment (djc) [reply quinn-proto/src/crypto.rs]: I think you're asking if we can mutate the existing `Crypto` instead of instantiating a new one? In which case, we cannot, because we need to keep the old `Crypto` around for older packets that might still come in.

**Comment by Ralith** [comment]
Another option might be to not keep `CryptoState`s around at all; store the secrets and IVs directly in the `CryptoContext, and construct the suitable AEAD key types. You'd probably want to tweak `CryptoState::update` to replace the `&self` argument with `secret: &[u8]` to make this convenient.

**Comment by stammw** [comment]
I've followed what `ZeroRttCrypto` seemed to work to move retained secrets of `OneRtt` into a new `OneRttCrypto` type. 

The remaining ZeroRtt code as been commented out as I'm not sure of how I should reshape it after removing `CryptoState`. Tell me if you want me to update the commented code.

I think it might be worth to writing a test for `Crypto::update()`. But for the moment it's far from being trivial for me...

File: quinn-proto/src/crypto.rs [hunk 5]
```diff
@@ -204,51 +247,58 @@ impl Crypto {
...
         use self::Crypto::*;
         match self {
-            Initial(ctx) | OneRtt(ctx) => ctx,
+            Initial(ctx) => ctx,
+            OneRtt(crypto) => &crypto.context,
         }
     }
 
     pub fn encrypt(&self, packet: u64, buf: &mut Vec<u8>, header_len: usize) {
         // FIXME: retain crypter
```
Comment (Ralith) [inline quinn-proto/src/crypto.rs]: Stale comment

File: quinn-proto/src/crypto.rs [hunk 6]
```diff
@@ -99,11 +99,16 @@ pub fn reset_token_for(key: &SigningKey, id: &ConnectionId) -> [u8; RESET_TOKEN_
     result
 }
 
-#[derive(Clone)]
+pub struct OneRttCrypto {
+    context: CryptoContext,
+    local_secret: Vec<u8>,
+    remote_secret: Vec<u8>,
+}
```
Comment (Ralith) [inline quinn-proto/src/crypto.rs]: I'm not sure this merits having a separate type, since it's harmless for initial to have these fields and either store its secrets in them for consistency, or just ignore them. @djc, do you have an opinion?

**Comment by Ralith** [comment]
I'd say you should feel free to nuke stale commented 0-RTT related code that the rest of the code has diverged from.

**Comment by stammw** [comment]
Ok great! Can I change the order of definitions a bit? I'm used to:

```
struct S1 { }
pub struct S2 { s1: S1 }

fn f1 { }

impl S1 {}
impl S2 { s1.foo() }
```

(When a symbol is used in the modules, it is always declare before hand.)

**Comment by stammw** [comment]
What about CookieFactory, should go away as well?

File: quinn-proto/src/crypto.rs (hunk 6 above)
Comment (djc) [reply quinn-proto/src/crypto.rs]: Yeah, I think it should just be a single type, and store the initial secrets in them.

**Comment by Ralith** [comment]
> Can I change the order of definitions a bit?

I prefer that associated items on a type always immediately follow the type definitions, with top-level items close to their use site. Refer to [the Rust style guidelines](https://doc.rust-lang.org/1.0.0/style/style/organization.html#define-types-and-operations-together.).

> What about CookieFactory, should it go away as well?

That's unrelated to your changes, so better to leave it alone. I should probably nuke it in #98 though, totally forgot it was there...

File: quinn-proto/src/crypto.rs [hunk 7]
```diff
@@ -133,25 +149,52 @@ impl Crypto {
...
+            local_pn_key,
+            local_iv,
+            opening_key: aead::OpeningKey::new(cipher, &remote_key).unwrap(),
+            remote_pn_key,
+            remote_iv,
             digest,
         })
     }
 
+    fn generate_1rtt(
```
Comment (djc) [inline quinn-proto/src/crypto.rs]: Can we call this `new_1rtt()` by analogy with `new_initial()`?

File: quinn-proto/src/crypto.rs [hunk 8]
```diff
@@ -102,7 +102,7 @@ pub fn reset_token_for(key: &SigningKey, id: &ConnectionId) -> [u8; RESET_TOKEN_
 pub enum Crypto {
```
Comment (djc) [inline quinn-proto/src/crypto.rs]: Actually, throughout this PR I have been feeling that this enum is not pulling its weight, and we should just use `CryptoContext` as the `Crypto` type. At this point, it seems we have more code working around the variants than using them, and I think usages in `Connection` are disambiguated by their usage sites. @Ralith what do you think?

@stammw I don't want to let your PR spiral out of control if you don't want it to, so do with this comment what you like.

**Comment by djc** [comment]
> Can I change the order of definitions a bit?

What @Ralith said, but also: over my learning Rust, I've come to appreciate that, because the compiler doesn't mandate that dependencies are defined before their dependees, for human readers it's often preferable to have high-level API up first and then fill in the details (dependencies) below that. So in your example, I'd order them like this:

```rust
pub struct S2 { s1: S1 }

impl S2 { s1.foo() }

struct S1 { }

impl S1 {}

fn f1 { }
```

(So, S2 comes before S1 because S2 depends on S1, structs and their impls close together, helper functions at the bottom.)

File: quinn-proto/src/crypto.rs (hunk 8 above)
Comment (Ralith) [reply quinn-proto/src/crypto.rs]: I'm 100% on board with ditching the enum outright.

**Comment by stammw** [comment]
@djc This is the structure I understood working on the file. I'll fix the codes I added in the wrong places then.

**Comment by stammw** [comment]
@djc This is the structure I understood working on the file. I'll fix the codes I added in the wrong places then.

File: quinn-proto/src/crypto.rs (hunk 6 above)
Comment (stammw) [reply quinn-proto/src/crypto.rs]: I'll move them to `CryptoContext` and assign them into `new_initial()` as well.

Then removing the `enum` should not be a problem anymore.

File: quinn-proto/src/crypto.rs [hunk 9]
```diff
@@ -200,54 +236,54 @@ impl Crypto {
...
     pub fn is_1rtt(&self) -> bool {
         match *self {
-            Crypto::OneRtt(_) => true,
+            Crypto::OneRtt(_, _, _) => true,
             _ => false,
         }
     }
 
-    pub fn write_nonce(&self, state: &CryptoState, number: u64, out: &mut [u8]) {
+    pub fn write_nonce(&self, iv: &Vec<u8>, number: u64, out: &mut [u8]) {
```
Comment (djc) [inline quinn-proto/src/crypto.rs]: There's almost never a good reason to pass a `&Vec<u8>`; use a `&[u8]` instead (on the caller side, it will coerce automatically).

File: quinn-proto/src/connection.rs [hunk 10]
```diff
@@ -2666,6 +2669,27 @@ impl Connection {
...
+        }
+    }
+
+    fn is_1rtt(&self) -> bool {
+        match *self {
+            CryptoLevel::OneRtt => true,
+            _ => false,
+        }
+    }
+}
```
Comment (Ralith) [inline quinn-proto/src/connection.rs]: Given that this enum likely won't ever contain data, these methods should be removed in favor of applying `#[derive(Eq, PartialEq)]` to the enum and inspecting values with `==`/`!=`.

Comment (stammw) [reply quinn-proto/src/connection.rs]: This is what I was looking for, thank you.

**Comment by stammw** [comment]
@Ralith Regarding updating keys, I thought that how they are rebuilt is defined in the RFC. But from what I understand, it is actually arbitrary and implementation-defined, as long as the updated keys are different from the previous ones? 

In this case using this kind of test cas should be sufficient:

```rust
let old = Crypto::new_1rtt();
let new = old.update();
assert(old.keys != new.keys)
```

Am I right? 

Couldn't I use `rand` crate to build them ?

**Comment by djc** [comment]
I've rebased and squashed these changes and submitted them for approval in #102. @stammw thanks for sticking with it!

**Comment by stammw** [comment]
Great! Thank you for the guidance. 

What about the unit test and the update "bug" here ? Should I continue with this?

Else, feel free to point what I can work on next !

**Comment by djc** [comment]
Yeah, continuing with that unit test and fixing the update code would be a great next step!
